core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java

```diff
@@ -0,0 +1,190 @@
+        this.index = index;
+        this.shard = shard;
+        this.primary = primary;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException validationException = null;
+        if (this.useAnyUnassignedShard() == false) {
+            if (this.index == null) {
+                validationException = addValidationError("index must be specified", validationException);
+            }
+            if (this.shard == null) {
+                validationException = addValidationError("shard must be specified", validationException);
+            }
+            if (this.primary == null) {
+                validationException = addValidationError("primary must be specified", validationException);
+            }
+        }
+        return validationException;
+    }
+
+    /**
+     * Returns {@code true} iff the first unassigned shard is to be used
+     */
+    public boolean useAnyUnassignedShard() {
+        return this.index == null && this.shard == null && this.primary == null;
+    }
+
+    public ClusterAllocationExplainRequest index(String index) {
+        this.index = index;
+        return this;
+    }
+
+    @Nullable
+    public String getIndex() {
+        return this.index;
+    }
+
+    public ClusterAllocationExplainRequest shard(Integer shard) {
+        this.shard = shard;
+        return this;
+    }
+
+    @Nullable
+    public int getShard() {
+        return this.shard;
+    }
+
+    public ClusterAllocationExplainRequest primary(Boolean primary) {
+        this.primary = primary;
+        return this;
+    }
+
+    @Nullable
+    public boolean isPrimary() {
+        return this.primary;
+    }
+
+    public void includeYesDecisions(boolean includeYesDecisions) {
+        this.includeYesDecisions = includeYesDecisions;
+    }
+
+    /** Returns true if all decisions should be included. Otherwise only "NO" and "THROTTLE" decisions are returned */
+    public boolean includeYesDecisions() {
+        return this.includeYesDecisions;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder sb = new StringBuilder("ClusterAllocationExplainRequest[");
+        if (this.useAnyUnassignedShard()) {
+            sb.append("useAnyUnassignedShard=true");
+        } else {
+            sb.append("index=").append(index);
+            sb.append(",shard=").append(shard);
+            sb.append(",primary?=").append(primary);
+        }
+        sb.append(",includeYesDecisions?=").append(includeYesDecisions);
+        return sb.toString();
+    }
+
+    public static ClusterAllocationExplainRequest parse(XContentParser parser) throws IOException {
+        String currentFieldName = null;
+        String index = null;
+        Integer shard = null;
+        Boolean primary = null;
+        XContentParser.Token token;
+        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```

Comment:
Sad!  Maybe fix it? Or we can split that to another PR if it is scary or
complicated.
On Mar 23, 2016 7:13 PM, "Lee Hinman" <notifications@github.com> wrote:

> In
> core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainRequest.java
> <https://github.com/elastic/elasticsearch/pull/17305#discussion_r57253131>
> :
>
> > +        } else {
> > +            sb.append("index=").append(index);
> > +            sb.append(",shard=").append(shard);
> > +            sb.append(",primary?=").append(primary);
> > +        }
> > +        sb.append(",includeYesDecisions?=").append(includeYesDecisions);
> > +        return sb.toString();
> > +    }
> > +
> > +    public static ClusterAllocationExplainRequest parse(XContentParser parser) throws IOException {
> > +        String currentFieldName = null;
> > +        String index = null;
> > +        Integer shard = null;
> > +        Boolean primary = null;
> > +        XContentParser.Token token;
> > +        while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
>
> Actually, had to revert that as ObjectParser doesn't correctly handle
> empty objects the way I want it to.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/17305/files/a5ac64ad0be0e841ba2d14689a9aad43c4e66e6b#r57253131>
>